pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java

```diff
@@ -543,7 +543,7 @@ private void parseSingleRuleNode(RuleSetReferenceId ruleSetReferenceId, RuleSetB
         if (attribute == null || "".equals(attribute)) {
             throw new IllegalArgumentException("The 'class' field of rule can't be null, nor empty.");
         }
-        Rule rule = (Rule) classLoader.loadClass(attribute).newInstance();
+        Rule rule = (Rule) RuleSetFactory.class.getClassLoader().loadClass(attribute).newInstance();
```

Comment:
This change in particular means users can no longer write custom Java rules within their project's codebase / auxclasspath and apply them. This feature was undocumented and unsupported.

Custom rules can be written, but must be part of PMD's execution classpath.